File: C++_Version/Hash_Table/463Island_Perimeter.cpp
//You are given a map in form of a two - dimensional integer grid where 1 represents land and 0 represents water.Grid cells are connected horizontally / vertically(not diagonally).The grid is completely surrounded by water, and there is exactly one island(i.e., one or more connected land cells).The island doesn't have "lakes" (water inside that isn't connected to the water around the island).One cell is a square with side length 1. The grid is rectangular, width and height don't exceed 100. Determine the perimeter of the island.

#include <vector>
#include <iostream>
#include <algorithm>
#include <cstdlib>
#include <functional>
#include <string>
#include <unordered_map>
#include <map>
using namespace std;

class Solution {
public:
	int islandPerimeter(vector<vector<int>>& grid) {
		int height = (int)grid.size();
		int width = (int)grid[0].size();

		int perimeter = 0;
		int temp;

		for (int i = 0; i < height; ++i)
		{
			for (int j = 0; j < width; ++j)
			{
				if (grid[i][j] == 1)
				{
					temp = position_detect(i, j, grid);
					perimeter += temp;
				}
				
			}
		}

		return perimeter;
	}

private:
	int position_detect(int i, int j, vector<vector<int>>& grid)
	{
		int temp = 0;

		if (i + 1 < (int)grid.size())
		{
			if (grid[i + 1][j] == 1)
				temp++;
		}
		if (i - 1 >= 0)
		{
			if (grid[i - 1][j] == 1)
				temp++;
		}
		if (j + 1 < (int)grid[0].size())
		{
			if (grid[i][j + 1] == 1)
				temp++;
		}
		if (j - 1 >= 0)
		{
			if (grid[i][j - 1] == 1)
				temp++;
		}

		return (4 - temp);
	}
};


int main()
{
	Solution sol1;
	string pattern = "abba", str = "dog cat cat dog";
	vector<int> a({ 1,3,2,2,5,2,3,7 });
	vector<vector<int>> grid({
		{0,1,0,0},
	{1,1,1,0},
	{0,1,0,0},
	{1,1,0,0} });

	int receive_flag = sol1.islandPerimeter(grid);

	system("pause");

	return 0;
}

File: Java_Version/463_Island_Perimeter.java
/*Author: Bochen (mddboc@foxmail.com)
Last Modified: Tue Apr 10 22:28:45 CST 2018*/

/*
You are given a map in form of a two-dimensional integer grid where 1 represents land and 0 represents water. Grid cells are connected horizontally/vertically (not diagonally). The grid is completely surrounded by water, and there is exactly one island (i.e., one or more connected land cells). The island doesn't have "lakes" (water inside that isn't connected to the water around the island). One cell is a square with side length 1. The grid is rectangular, width and height don't exceed 100. Determine the perimeter of the island.

        Example:

        [[0,1,0,0],
        [1,1,1,0],
        [0,1,0,0],
        [1,1,0,0]]

        Answer: 16
        Explanation: The perimeter is the 16 yellow stripes in the image below:
*/


import java.util.*;
import java.lang.Math;
import java.lang.System;
import java.lang.Integer;


public class Main {

    public static void main(String[] args) throws ArithmeticException {

        int a = 3, b = 0;


        int result = a / b;


        System.out.println("haha");
    }

}


class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
    }
}


class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}


class Solution {
    public int islandPerimeter(int[][] grid) {

        int rows = grid.length;
        int cols = grid[0].length;

        int gridCount = 0;
        int leftOrUpRepeatCount = 0;

        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < cols; col++) {

                if (grid[row][col] == 1) {
                    gridCount++;

                    if (row != 0 && grid[row - 1][col] == 1) {
                        leftOrUpRepeatCount++;
                    }
                    if (col != 0 && grid[row][col - 1] == 1) {
                        leftOrUpRepeatCount++;
                    }
                }
            }
        }

        return (gridCount * 4 - leftOrUpRepeatCount * 2);
    }
}